src/Oauth2Client.php

```diff
@@ -0,0 +1,358 @@
+  {
+    $required = array('file', 'fileName', 'hasRightsToModel', 'acceptTermsAndConditions');
+    foreach ($required as $key) {
+      if (!array_key_exists($key, $params)) {
+        throw new ParameterValidationException('Shapeways\Oauth2CurlClient::uploadModel missing required key: ' . $key);
+      }
+    }
+
+    $params['file'] = rawurlencode(base64_encode($params['file']));
+
+    $url = $this->baseApiUrl . '/models/' . $this->apiVersion;
+    return $this->_post($url, $params,
+      array('Authorization' => 'Bearer ' . $this->accessToken, 'Content-type' => 'application/json'));
+  }
+
+  /**
+   * Get information for the provided modelId
+   *
+   * @link https://developers.shapeways.com/api-reference#Models
+   *
+   * @param int $modelId the modelId for the model to retreive
+   * @return array the json response from the api call
+   */
+  public function getModelInfo($modelId)
+  {
+    $url = $this->baseApiUrl . '/models/' . $modelId . '/' . $this->apiVersion;
+
+    return $this->_get($url);
+  }
+
+
+  /**
+   * Get a list of materials
+   *
+   * @link https://developers.shapeways.com/api-reference#Materials
+   *
+   * @return array the json response from the api call
+   */
+  public function getMaterials()
+  {
+    $url = $this->baseApiUrl . '/materials/' . $this->apiVersion;
+
+    return $this->_get($url);
+  }
+
+  /**
+   * place an order for the user
+   *
+   * @link https://developers.shapeways.com/api-reference#Order
+   *
+   * @param array $params the order data to set
+   * @return array the json response from the api call
+   * @throws ParameterValidationException
+   */
+  public function placeOrder($params)
+  {
+    $required = array(
+      'items',
+      'firstName',
+      'lastName',
+      'country',
+      'state',
+      'city',
+      'address1',
+      'zipCode',
+      'phoneNumber',
+      'shippingOption'
+    );
+
+    foreach ($required as $key) {
+      if (!array_key_exists($key, $params)) {
+        throw new ParameterValidationException('Shapeways\Oauth2CurlClient::placeOrder missing required key: ' . $key);
+      }
+    }
+
+    $url = $this->baseApiUrl . '/orders/' . $this->apiVersion;
+    return $this->_post($url, $params,
+      array('Authorization' => 'Bearer ' . $this->accessToken, 'Content-type' => 'application/json'));
+  }
+
+
+  /**
+   * Get information for the provided orderId
+   *
+   * @link https://developers.shapeways.com/api-reference#Order
+   *
+   * @param int $oderId the orderId for the Order to retreive
+   * @return array the json response from the api call
+   */
+  public function getOrderInfo($oderId)
+  {
+    $url = $this->baseApiUrl . '/orders/' . $oderId . '/' . $this->apiVersion;
+    return $this->_get($url);
+  }
+
+  /**
+   * @param $url
+   * @param array $params
+   * @param array $headers
+   * @param array $auth
+   * @return mixed
+   */
+  private function _post($url, $params = array(), $headers = array(), $auth = array()) {
```

Comment:
I do not like underscore methods - those are like almost internal reserved PHP keywords and functions. It's private, isn't it nicer to call it `postRequest()`?

Reply:
also, for the fancier result you can have `makeRequest` method and pass parameter "get" or "post" into it, and abstract code to have a separate request caller. this will prevent all the code repetition


something like this? (you can ignore middleware)
https://github.com/shapeways/portal/blob/master/www/roadrunner/Infrastructure/RequestHelper.php#L39

Reply:
surething :) updated